### ISSUE TITLE
Slow infererence speed with the binding

### ISSUE THREAD

**Issue by phamson02** [namtranase/gemma-cpp-python#12]
With the same prompt, gemma.cpp's interactive mode can complete the answer in ~2s while gemma-cpp-python takes >1 min. This might be due to the GIL as shown in the CPU's cores usage in two screenshots below.

(1) gemma-cpp-binding
![gemma-cpp-binding](https://github.com/namtranase/gemma-cpp-python/assets/82074941/78b7e2df-8ab0-4612-bdc0-215891780388)

(2) gemma.cpp
![Screenshot from 2024-03-25 12-14-59](https://github.com/namtranase/gemma-cpp-python/assets/82074941/800894d9-2173-446f-808e-f04d05a130ed)

Might related to this issue:
[1] [How to use pybind11 in multithreaded application](https://stackoverflow.com/questions/47309688/how-to-use-pybind11-in-multithreaded-application)
[2] [pybind11's doc](https://pybind11.readthedocs.io/en/stable/advanced/misc.html#global-interpreter-lock-gil)

**Comment by namtranase** [comment]
Thank @phamson02 for pointing it out, I think we can add logs and make sure what are steps that make pybind module slow. 
And also test cases.

**Comment by phamson02** [comment]
@namtranase Can you check if this is also a problem on your end?

I tried the lastest build on Windows (on Python 3.12 and 3.11) and didn't encounter this. But on my Ubuntu, the performance issue still happens on Python 3.12 and 3.11, only one core are utilized.

**Comment by namtranase** [comment]
Yes Son I will check it on my environments.

**Comment by omkar806** [comment]
@phamson02  is this issue resolved ?

**Comment by phamson02** [comment]
> @phamson02 is this issue resolved ?

Yes, it was fixed by #13 and the rewritten in #15 also took this into account. The issue reported was due to threads being pinned during model loading without being released, resulting in only a few cores being utilized for inference. And I think it was specific to Linux machines with > 12 CPU cores.
